File: src/client/shared/modelCatalog.ts
import catalogData from "../../shared/model_catalog.json";

export interface ModelOption {
  value: string;
  label: string;
}

type Catalog = {
  models: Record<string, ModelOption>;
  groups: Record<string, string[]>;
  defaults: Record<string, string>;
};

const catalog = catalogData as Catalog;

export function modelOption(modelKey: string): ModelOption {
  const model = catalog.models[modelKey];
  if (!model) throw new Error(`Unknown model key: ${modelKey}`);
  return { value: model.value, label: model.label };
}

export function modelOptions(group: string): ModelOption[] {
  const keys = catalog.groups[group];
  if (!keys) throw new Error(`Unknown model group: ${group}`);
  return keys.map(modelOption);
}

export function defaultModel(role: string): string {
  const modelKey = catalog.defaults[role];
  if (!modelKey) throw new Error(`Unknown default model role: ${role}`);
  return modelOption(modelKey).value;
}

export const LLM_MODELS = modelOptions("llm");
export const AGENT_MODELS = modelOptions("agent");
export const TINKER_MODELS = modelOptions("tinker");

export const DEFAULT_LLM_MODEL = defaultModel("llm");
export const DEFAULT_AGENT_MODEL = defaultModel("agent");
export const DEFAULT_TINKER_MODEL = defaultModel("tinker");

File: src/shared/model_catalog.py
from __future__ import annotations

import json
from functools import cache
from pathlib import Path
from typing import Any


@cache
def catalog() -> dict[str, Any]:
    return json.loads((Path(__file__).with_name("model_catalog.json")).read_text())


def model_option(model_key: str) -> dict[str, str]:
    try:
        model = catalog()["models"][model_key]
    except KeyError as exc:
        raise KeyError(f"Unknown model key: {model_key}") from exc
    return {"value": model["value"], "label": model["label"]}


def model_options(group: str) -> list[dict[str, str]]:
    try:
        keys = catalog()["groups"][group]
    except KeyError as exc:
        raise KeyError(f"Unknown model group: {group}") from exc
    return [model_option(key) for key in keys]


def model_values(group: str) -> list[str]:
    return [option["value"] for option in model_options(group)]


def default_model(role: str) -> str:
    try:
        model_key = catalog()["defaults"][role]
    except KeyError as exc:
        raise KeyError(f"Unknown default model role: {role}") from exc
    return model_option(model_key)["value"]
